website/blog/2024-06-26-llm-as-judge/index.md

```diff
@@ -0,0 +1,522 @@
+
+start by installing all the necessary libraries for this demo to work.
+Since mlflow evaluate offers some default metrics we need to install the libraries to calculate those metrics (toxiticy etc..)
+
+```python
+!pip install mlflow==2.14.1 openai  transformers torch torchvision evaluate datasets openai tiktoken fastapi rouge_score textstat tenacity plotly ipykernel nbformat==5.10.4
+
+!pip install --upgrade nbformat
+```
+
+
+We will be using gpt3.5 and gpt4 during this example for that let's start by setting up an open ai key
+```python
+%env OPENAI_API_KEY=<your-api-key>
+```
+
+Import the necessary libraries
+```python
+import mlflow
+import os
+# Run a quick validation that we have an entry for the OPEN_API_KEY within environment variables
+assert "OPENAI_API_KEY" in os.environ, "OPENAI_API_KEY environment variable must be set"
+
+import openai
+import pandas as pd
+```
+
+When using mlflow.evaluate(), your LLM can be either:
+1.	A `mlflow.pyfunc.PyFuncModel()` . Generally speaking an MLflow model. 
+2.	A `python function` that has strings as inputs and outputs a single string. 
+3.	An `MLflow Deployments` endpoint URI.
+4.	Set `model=None` when you only want to provide the data which has already been scored by the model.
+
+For this case let's use an mlflow model.
+
+Let's start by logging our translation model on mlflow.
+For this tutorial let's use a gpt3.5 with a system prompt
+
+```python
+
+system_prompt = "Translate the following sentences into Spanish"
+basic_translation_model = mlflow.openai.log_model(
+    model="gpt-3.5-turbo",
+    task=openai.chat.completions,
+    artifact_path="model",
+    messages=[
+        {"role": "system", "content": system_prompt},
+        {"role": "user", "content": "{user_input}"},
+    ],
+)
+```
+
+Testing the model to make sure it works
+
+```python
+model = mlflow.pyfunc.load_model(basic_translation_model.model_uri)
+
+model.predict("Hello, how are you?")
+
+# Output = ['¡Hola, ¿cómo estás?']
+```
+
+To use mlflow evaluate we need to prepare some data to provide as input to our LLM.
+In this example we will define some common english expressions which we want translated.
+
+```python
+# 1. Prepare evaluation data
+eval_data = pd.DataFrame(
+    {
+        "llm_inputs": [
+            "I'm over the moon about the news!",
+            "Spill the beans.",
+            "Bite the bullet.",
+            "Better late than never.",
+            
+        ]
+    }
+)
+```
+
+
+This example demonstrates how to create a custom metric that uses an LLM to judge the cultural sensitivity of translations. The LLM considers context, idiomatic expressions, and cultural nuances that might be missed by traditional metrics.
+
+To achieve this let's define some custom metric. 
+We want to adress how faithfull the translation is, for that we have to consider cultural factors and not only a literal translation.
+
+Let's set a metric that takes into account that cultural sensitivity.
```

Comment:
little more detail here—what does it mean to account for cultural sensitivity? What is the purpose of the metric in the context of the example company? (see comments in first section as well, no need to repeat it if you add more about this point above)

Reply:
I've added some examples in the intro section of a poorly translated comment and how it would affect the score and impact the company, let me know if it fits what you had in mind